doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,271 @@
+being assigned to data from upstream sources. All reads are given a timestamp equal to the global timestamp. All writes
+to user tables are given a timestamp larger than the global timestamp, and when the write completes then the timestamp
+should be advanced to the timestamp of that write.
+
+As an example, a timeline's timestamp can be initialized to the current system time and be advanced periodically to the
+current value of a monotonically increasing system clock. Another example is that a timeline's timestamps can be
+initialized to the timestamp's minimum value and can be periodically advanced in the following
+way: `global_timestamp = max(min(uppers) - 1, global_timestamp)`, where `uppers` is the list of all `upper`s in the
+timeline. These are just potential implementations and not the only correct implementations.
+
+If some operation is given timestamp `ts`, then once that operation completes the global timestamp must be larger than
+or equal to `ts`. How this specific property is achieved for reads is described
+in [Read Capabilities on Global Timestamp](#Read Capabilities on Global Timestamp) and how it's achieved for writes to
+user tables is described in [Group Commit/Write Coalesce](#Group Commit/Write Coalesce).
+
+The properties described in this section are sufficient to ensure Strict Serializability across one start of
+Materialize (i.e. not between restarts).
+
+- [X] All reads and writes occur at specific timestamps: satisfied by the Coordinator assigning timestamps to all
+  operations.
+- [X] These timestamps are non-decreasing in real-time order: all timestamps are assigned sequentially on a single
+  thread. Each timestamp is guaranteed to be larger than or equal to the timestamp of the most recently completed query.
+- [X] With strict increases for writes following reads: the timestamp assigned to all writes is larger than all previous
+  timestamps (See [Group Commit/Write Coalesce](#Group Commit/Write Coalesce)).
+- [X] Each timestamp is assigned to an event at some real-time moment within the event's bounds: For any two events e1
+  and e2, if e2 starts after e1 finished, then e2's timestamp will be larger or equal timestamp to e1's timestamp.
+- [X] Each timestamp transition is made durable before any subsequent response is issued: writes wait for an `append`
+  command to be made durable before returning a response to the client, additionally timestamp transitions are persisted
+  to disk (See [Global Timestamp Recovery](#Global Timestamp Recovery)).
+
+### Global Timestamp Recovery
+
+After a restart the Coordinator must reestablish the global timestamp to some value greater than or equal to the value
+of the previous global timestamp before the restart. This ensures that a read or write after the restart is assigned a
+timestamp greater than or equal to all reads and writes before the restart.
+
+Proposal: Use a [Percolator](https://storage.googleapis.com/pub-tools-public-publication-data/pdf/36726.pdf) inspired
```

Comment:
It seems like if we use the largest timestamp persisted for any write, then reads need to block on writes being made durable, so that we don't report a result at `time+1` and then restart and report a result at `time`, which would be a linearizability violation. The percolator approach, as I understand it, would allow us to service reads without needing to make anything durable before we do this (unless the read depends on the thing we might make durable).

Reply:
I think the following scenario would be a violation if we just used the most recent write

1. Timestamp is at `ts` (a write just completed at `ts).
2. Table advancement bumps the timestamp to `ts + 100` and queues a message for the Coord.
4. Someone reads from a view over an upstream source at `ts + 100`.
5. The read completes and is returned to the client.
6. The Coord crashes and restarts.
7. The timestamp is re-established to `ts`.
8. Someone performs a read at `ts`.

Reply:
Step 2 says it bumps and queues a message. Are we able to only bump when the message is processed instead of when the message is queued? The group commit stuff in 12979 enqueues a request to bump, but does the actual bumping in the group commit step.

Reply:
That should be possible. The one issue is that if we include DDL, then what I described here is no longer true: https://github.com/MaterializeInc/materialize/pull/12979#discussion_r900466269. You would be able to read system tables before they're updated and see a discrepancy with the in-memory state.

Though waiting until to bump the timestamp until the message in processed gives us the benefit of being able to read from tables while we wait for group commit. So it might be worth figuring out how to do.

Reply:
After thinking about this for a couple of days, I think your suggestion is better and we should go back to the old recovery method here.